cudf/cudaarray.py

```diff
@@ -0,0 +1,282 @@
+        shape = shape,
+    elif shape == ():
+        shape = (1,)
+    if isinstance(strides, integer_types):
+        strides = strides,
+    else:
+        if strides is None:
+            nd = len(shape)
+            strides = [0] * nd
+            if order == 'C':
+                strides[-1] = dtype.itemsize
+                for d in reversed(range(nd - 1)):
+                    strides[d] = strides[d + 1] * shape[d + 1]
+            elif order == 'F':
+                strides[0] = dtype.itemsize
+                for d in range(1, nd):
+                    strides[d] = strides[d - 1] * shape[d - 1]
+            else:
+                raise ValueError('order must be "C" or "F" but got %r'
+                                 % (order))
+    assert len(strides) == len(shape)
+    return tuple(shape), tuple(strides), dtype
+
+
+class CudaNDArray(object):
+    """CudaNDArray represents CUDA device memory as a multidimensional
+    array that items can be accessed from host.
+
+    The device memory is accessed via pyarrow.cuda.CudaBuffer instance.
+
+    numba.cuda defines DeviceNDArray for the same purpose as
+    CudaNDArray but DeviceNDArray accesses the device memory via
+    device pointer (stored in MemoryPointer). Since numba jit
+    functions use DeviceNDArray, we implement cooperativity between
+    the two representations.
+
+    """
+
+    def __init__(self, shape, strides=None, dtype=None, cuda_data=None,
+                 order='C', ctx=None):
+        self.shape, self.strides, self.dtype \
+            = prepare_shape_strides_dtype(shape, strides, dtype, order=order)
+        self.ndim = len(self.shape)
+        self.size = int(np.prod(self.shape))
+        self.alloc_size = self.size * self.dtype.itemsize
+        self._dummy = dummyarray.Array(self.shape, strides=self.strides,
+                                       dtype=self.dtype, order=order)
+        if cuda_data is None:
+            if ctx is None:
+                ctx = pa.cuda.Context()
+            if not isinstance(ctx, pa.cuda.Context):
+                ctx = pa.cuda.Context.from_numba(ctx)
+            cuda_data = ctx.new_buffer(self.alloc_size)
+        elif isinstance(cuda_data, pa.cuda.CudaBuffer):
+            pass
+        elif isinstance(cuda_data, CudaNDArray):
+            raise NotImplementedError('CUDA device array from a device array')
+        else:
+            if ctx is None:
+                ctx = pa.cuda.Context()
+            if not isinstance(ctx, pa.cuda.Context):
+                ctx = pa.cuda.Context.from_numba(ctx)
+
+            if cuda_data.dtype.char == 'M':
+                cuda_data = cuda_data.view('uint64')
+            cuda_data = ctx.buffer_from_data(cuda_data, size=self.alloc_size)
+
+        self.cuda_data = cuda_data  # CudaBuffer
+
+    def __repr__(self):
+        return (f'{type(self).__name__}({self.shape}, strides={self.strides},'
+                f'dtype={self.dtype}, cuda_data={self.cuda_data})')
```

Comment:
@kkraus14 - what is the minimum version of Python that cudf should support?

Reply:
Fixed.